src/main/java/org/hydev/mcpm/client/database/searchusecase/SearcherByKeyword.java

```diff
@@ -0,0 +1,77 @@
+package org.hydev.mcpm.client.database.searchusecase;
+
+import org.hydev.mcpm.client.models.PluginModel;
+
+import java.util.*;
+
+/**
+ * Searcher that returns a map based on keywords.
+ *
+ * @author Jerry Zhu (<a href="https://github.com/jerryzhu509">...</a>)
+ */
+public class SearcherByKeyword implements Searcher {
+
+    private static Map<String, List<PluginModel>> keywordMap = null;
+
+    /**
+     * Returns a dictionary mapping the different keywords to the matching plugins
+     * This function should ignore letter case in searching.
+     * For example, searching "java" would match "Java" as well.
+     * This function implements fuzzy search that doesn't require the exact phrase to be available,
+     * but requires all words in the phrase to be present.
+     * For example, searching "java 11" would match "java jdk 11" but not "java"
+     *
+     * @param plugins A list of all plugins in the database.
+     * @return A dictionary associating each keyword to the matching plugins.
+     */
+    @Override
+    public Map<String, List<PluginModel>> constructSearchMaps(List<PluginModel> plugins) {
+        Map<String, List<PluginModel>> models = new HashMap<>();
+        for (PluginModel plugin : plugins) {
+            // Get latest version
+            var v = plugin.getLatestPluginVersion();
+            if (v.isPresent() && v.get().meta() != null &&
+                    v.get().meta().description() != null && !v.get().meta().description().equals("")) {
+                String[] keywords = v.get().meta().description()
+                        .replaceAll("[^a-zA-Z ]", "").toLowerCase().split("\\s+");
+                for (String keyword : keywords) {
+                    if (!models.containsKey(keyword))
+                        models.put(keyword, new ArrayList<>());
+                    models.get(keyword).add(plugin);
+                }
+            }
+        }
+        return models;
+    }
+
+    /**
+     * Searches for plugins based on the provided user input.
+     *
+     * @param inp User input for the search. Should be a name as a non-empty string here.
+     * @param plugins A list of all plugins in the database.
+     * @return A dictionary associating a string feature of the plugins to the matching plugins.
+     */
+    @Override
+    public List<PluginModel> getSearchList(Object inp, List<PluginModel> plugins) {
+
+        // Instantiate if null
+        if (SearcherByKeyword.keywordMap == null) {
+            SearcherByKeyword.keywordMap = constructSearchMaps(plugins);
+        }
+
+        String [] keywords = ((String) inp).toLowerCase().split(" "); // Should be a string
+        Set<PluginModel> res = new HashSet<>();
+        boolean first = true;
+        for (String keyword : keywords) {
+            List<PluginModel> pl = SearcherByKeyword.keywordMap.get(keyword);
+            if (first) {
```

Comment:
Good job implementing it!
I noticed a few simplifications we can do. 
You noticed there is no universal set, so the first iteration has to add instead of intersect. 
But observe that doing this on the first iteration only is the same as doing this before the loop. This removes the if statement. 
But if you remove the addAll and replace it with an assignment, you’ll end up modifying the set, which modifies the index. However instead of using addAll on a new set you can use the set constructor directly to copy all the elements from the indexing set. 
Also the string cast is sus